File: app/src/main/java/socioverse/tifor/Activity/MainActivity.java
package socioverse.tifor.Activity;

import android.app.Application;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.AudioAttributes;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.Base64;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.firebase.firestore.DocumentChange;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.messaging.FirebaseMessaging;
import com.zegocloud.uikit.prebuilt.call.config.ZegoNotificationConfig;
import com.zegocloud.uikit.prebuilt.call.invite.ZegoUIKitPrebuiltCallInvitationConfig;
import com.zegocloud.uikit.prebuilt.call.invite.ZegoUIKitPrebuiltCallInvitationService;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import socioverse.tifor.Adapter.RecentConversionAdapter;
import socioverse.tifor.Model.ChatMessage;
import socioverse.tifor.Model.User;
import socioverse.tifor.R;
import socioverse.tifor.databinding.ActivityMainBinding;
import socioverse.tifor.listeners.ConversionListener;
import socioverse.tifor.utlities.Constants;
import socioverse.tifor.utlities.PreferenceManager;

public class MainActivity extends BaseActivity2 implements ConversionListener {

    private ActivityMainBinding binding;
    private PreferenceManager preferenceManager;
    private List<ChatMessage> conversions;
    private RecentConversionAdapter conversionAdapter;
    private FirebaseFirestore database;
    private final Boolean isReceiverAvailable = false;
    private User user;
    private final String channelname = "Socioverse";
    private final String channelId = "socioverse";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        try {
            preferenceManager = new PreferenceManager(getApplicationContext());
            loadUserDetails();
            init();
            getToken();
            setListeners();
            listenConversation();
            showNotification();
            startService();

            // Enable system default mode
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);

        } catch (Exception e) {

        }

        binding.imageProfile.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                try {
                    Intent intent = new Intent(getApplicationContext(), ProfileActivity.class);
                    startActivity(intent);
                } catch (Exception e) {

                }

            }
        });

    }

    private void init() {

        try {
            conversions = new ArrayList<>();
            conversionAdapter = new RecentConversionAdapter(conversions, this);
            binding.conversionRecyclerView.setAdapter(conversionAdapter);
            database = FirebaseFirestore.getInstance();
        } catch (Exception e) {

        }
    }

    private void setListeners() {

        try {
            binding.imageSignOut.setOnClickListener(v -> logOut());
            binding.fabNewChat.setOnClickListener(v -> startActivity(new Intent(getApplicationContext(), UsersActivity.class)));
        } catch (Exception e) {

        }
    }


    private void loadUserDetails() {


        try {
            binding.textName.setText(preferenceManager.getString(Constants.KEY_NAME));
            String encodedString = preferenceManager.getString(Constants.KEY_IMAGE);
            if (encodedString != null) {
                byte[] bytes = Base64.decode(encodedString, Base64.DEFAULT);
                Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
                binding.imageProfile.setImageBitmap(bitmap);
            }
        } catch (Exception e) {

        }

    }

    private void showToast(String message) {

        try {
            Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT).show();
        } catch (Exception e) {

        }

    }

    private void listenConversation() {

        try {
            database.collection(Constants.KEY_COLLECTION_CONVERSATIONS).whereEqualTo(Constants.KEY_SENDER_ID, preferenceManager.getString(Constants.KEY_USER_ID)).addSnapshotListener(eventListener);
            database.collection(Constants.KEY_COLLECTION_CONVERSATIONS).whereEqualTo(Constants.KEY_RECEIVER_ID, preferenceManager.getString(Constants.KEY_USER_ID)).addSnapshotListener(eventListener);

        } catch (Exception e) {

        }
    }


    private final EventListener<QuerySnapshot> eventListener = (value, error) -> {

        try {
            if (error != null) {

                return;

            }

            if (value != null) {

                for (DocumentChange documentChange : value.getDocumentChanges()) {

                    if (documentChange.getType() == DocumentChange.Type.ADDED) {

                        String senderId = documentChange.getDocument().getString(Constants.KEY_SENDER_ID);
                        String receiverId = documentChange.getDocument().getString(Constants.KEY_RECEIVER_ID);
                        ChatMessage chatMessage = new ChatMessage();
                        chatMessage.senderId = senderId;
                        chatMessage.receiverId = receiverId;
                        if (preferenceManager.getString(Constants.KEY_USER_ID).equals(senderId)) {

                            chatMessage.conversionImage = documentChange.getDocument().getString(Constants.KEY_RECEIVER_IMAGE);
                            chatMessage.conversionName = documentChange.getDocument().getString(Constants.KEY_RECEIVER_NAME);
                            chatMessage.conversionId = documentChange.getDocument().getString(Constants.KEY_RECEIVER_ID);

                        } else {

                            chatMessage.conversionImage = documentChange.getDocument().getString(Constants.KEY_SENDER_IMAGE);
                            chatMessage.conversionName = documentChange.getDocument().getString(Constants.KEY_SENDER_NAME);
                            chatMessage.conversionId = documentChange.getDocument().getString(Constants.KEY_SENDER_ID);

                        }
                        chatMessage.message = documentChange.getDocument().getString(Constants.KEY_LAST_MESSAGE);
                        chatMessage.dateObject = documentChange.getDocument().getDate(Constants.KEY_TIMESTAMP);
                        conversions.add(chatMessage);

                    } else if (documentChange.getType() == DocumentChange.Type.MODIFIED) {

                        for (int i = 0; i < conversions.size(); i++) {

                            String senderId = documentChange.getDocument().getString(Constants.KEY_SENDER_ID);
                            String receiverId = documentChange.getDocument().getString(Constants.KEY_RECEIVER_ID);
                            if (conversions.get(i).senderId.equals(senderId) && conversions.get(i).receiverId.equals(receiverId)) {

                                conversions.get(i).message = documentChange.getDocument().getString(Constants.KEY_LAST_MESSAGE);
                                conversions.get(i).dateObject = documentChange.getDocument().getDate(Constants.KEY_TIMESTAMP);
                                break;

                            }

                        }

                    }

                }

                Collections.sort(conversions, (obj1, obj2) -> obj2.dateObject.compareTo(obj1.dateObject));
                conversionAdapter.notifyDataSetChanged();
                binding.conversionRecyclerView.smoothScrollToPosition(0);
                binding.conversionRecyclerView.setVisibility(View.VISIBLE);
                binding.progressBar.setVisibility(View.GONE);
            }
        } catch (Exception e) {

        }

    };

    private void getToken() {

        try {
            FirebaseMessaging.getInstance().getToken().addOnSuccessListener(this::updateToken);
        } catch (Exception e) {

        }

    }

    private void updateToken(String token) {

        try {
            preferenceManager.putString(Constants.KEY_FCM_TOKEN, token);
            FirebaseFirestore database = FirebaseFirestore.getInstance();
            DocumentReference documentReference = database.collection(Constants.KEY_COLLECTION_USERS).document(preferenceManager.getString(Constants.KEY_USER_ID));
            documentReference.update(Constants.KEY_FCM_TOKEN, token).addOnFailureListener(e -> showToast("Unable To Update Token"));
        } catch (Exception e) {

        }
    }

    private void logOut() {
        try {
            showToast("LogOut..."); // Assuming this is a method to show a toast message
            FirebaseFirestore database = FirebaseFirestore.getInstance();
            DocumentReference documentReference = database.collection(Constants.KEY_COLLECTION_USERS).document(preferenceManager.getString(Constants.KEY_USER_ID));

            HashMap<String, Object> update = new HashMap<>();
            update.put(Constants.KEY_FCM_TOKEN, FieldValue.delete());

            documentReference.update(update).addOnSuccessListener(unused -> {
                // Clear preferences
                preferenceManager.clear();
                startActivity(new Intent(getApplicationContext(), LogInActivity.class));
                finish();

            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {
                    showToast("Faiuld LogOut");
                }
            });
        } catch (Exception e) {
            // Handle exceptions
        }
    }

    @Override
    public void OnConversionClicked(User user) {

        try {
            Intent intent = new Intent(getApplicationContext(), ChatActivity.class);
            intent.putExtra(Constants.KEY_USER, user);
            startActivity(intent);
        } catch (Exception e) {

        }

    }

    private void showNotification() {
        try {

            Uri customSoundUri = Uri.parse("android.resource://" + getPackageName() + "/" + R.raw.notification);

            NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

            // Since android Oreo notification channel is needed.
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {

                // Set the custom sound for the notification.
                AudioAttributes audioAttributes = new AudioAttributes.Builder().setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION).setUsage(AudioAttributes.USAGE_NOTIFICATION).build();

                NotificationChannel channel = new NotificationChannel(channelId, channelname, NotificationManager.IMPORTANCE_HIGH);
                channel.setSound(customSoundUri, audioAttributes);
                channel.setShowBadge(true);
                long[] vibrationPattern = {100, 50, 100, 50, 100, 50, 100, 50, 100};
                channel.setVibrationPattern(vibrationPattern);

                notificationManager.createNotificationChannel(channel);
            }


        } catch (Exception e) {

        }

    }

    private void startService() {

        try {
            Application application = getApplication(); // Android's application context
            long appID = 900851183;   // yourAppID
            String appSign = "e2dbd39a96fcba5c14649d787451893020a522a12891490ac92bd90e2a62ad3f";  // yourAppSign
            String userID = preferenceManager.getString(Constants.KEY_EMAIL); // yourUserID, userID should only contain numbers, English characters, and '_'.
            String userName = preferenceManager.getString(Constants.KEY_NAME);   // yourUserName

            ZegoUIKitPrebuiltCallInvitationConfig callInvitationConfig = new ZegoUIKitPrebuiltCallInvitationConfig();
            callInvitationConfig.notifyWhenAppRunningInBackgroundOrQuit = true;
            ZegoNotificationConfig notificationConfig = new ZegoNotificationConfig();
            notificationConfig.sound = "zego_uikit_sound_call";
            notificationConfig.channelID = "CallInvitation";
            notificationConfig.channelName = "CallInvitation";
            ZegoUIKitPrebuiltCallInvitationService.init(getApplication(), appID, appSign, userID, userName, callInvitationConfig);
        } catch (Exception e) {

        }

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        ZegoUIKitPrebuiltCallInvitationService.unInit();
    }

    @Override
    protected void onStart() {
        super.onStart();
        try {
            loadUserDetails();
            getToken();
            showNotification();
            startService();
        } catch (Exception e) {

        }
    }
}

File: app/src/main/java/socioverse/tifor/Activity/MainActivity.kt
package socioverse.tifor.Activity

import android.Manifest
import android.annotation.SuppressLint
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.media.AudioAttributes
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Base64
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.ActivityCompat
import com.google.firebase.firestore.DocumentChange
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.QuerySnapshot
import com.google.firebase.messaging.FirebaseMessaging
import socioverse.tifor.Adapter.RecentConversionAdapter
import socioverse.tifor.Model.ChatMessage
import socioverse.tifor.Model.User
import socioverse.tifor.R
import socioverse.tifor.databinding.ActivityMainBinding
import socioverse.tifor.listeners.ConversionListener
import socioverse.tifor.utlities.Constants
import socioverse.tifor.utlities.PreferenceManager
import java.util.Collections

class MainActivity : BaseActivity2(), ConversionListener {
    private var binding: ActivityMainBinding? = null
    private var preferenceManager: PreferenceManager? = null
    private var conversions: MutableList<ChatMessage>? = null
    private var conversionAdapter: RecentConversionAdapter? = null
    private var database: FirebaseFirestore? = null
    private val user: User? = null
    private val channelname = "Socioverse"
    private val channelId = "socioverse"
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        try {
            preferenceManager = PreferenceManager(applicationContext)
            loadUserDetails()
            init()
            token
            setListeners()
            listenConversation()
            permissionNotification()
            showNotification()

            // Enable system default mode
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
        } catch (e: Exception) {
        }
        binding!!.imageProfile.setOnClickListener {
            try {
                val intent = Intent(applicationContext, ProfileActivity::class.java)
                startActivity(intent)
            } catch (e: Exception) {
            }
        }
    }

    private fun init() {
        try {
            conversions = ArrayList()
            conversionAdapter = RecentConversionAdapter(conversions as ArrayList<ChatMessage>, this)
            binding!!.conversionRecyclerView.adapter = conversionAdapter
            database = FirebaseFirestore.getInstance()
        } catch (e: Exception) {
        }
    }

    private fun setListeners() {
        try {
            binding!!.imageSignOut.setOnClickListener { v: View? -> logOut() }
            binding!!.fabNewChat.setOnClickListener { v: View? ->
                startActivity(
                    Intent(
                        applicationContext, UsersActivity::class.java
                    )
                )
            }
        } catch (e: Exception) {
        }
    }

    private fun loadUserDetails() {
        try {
            binding!!.textName.text = preferenceManager!!.getString(Constants.KEY_NAME)
            val encodedString = preferenceManager!!.getString(Constants.KEY_IMAGE)
            if (encodedString != null) {
                val bytes = Base64.decode(encodedString, Base64.DEFAULT)
                val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                binding!!.imageProfile.setImageBitmap(bitmap)
            }
        } catch (e: Exception) {
        }
    }

    private fun showToast(message: String) {
        try {
            Toast.makeText(applicationContext, message, Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
        }
    }

    private fun listenConversation() {
        try {
            database!!.collection(Constants.KEY_COLLECTION_CONVERSATIONS).whereEqualTo(
                Constants.KEY_SENDER_ID, preferenceManager!!.getString(
                    Constants.KEY_USER_ID
                )
            ).addSnapshotListener(eventListener)
            database!!.collection(Constants.KEY_COLLECTION_CONVERSATIONS).whereEqualTo(
                Constants.KEY_RECEIVER_ID, preferenceManager!!.getString(
                    Constants.KEY_USER_ID
                )
            ).addSnapshotListener(eventListener)
        } catch (e: Exception) {
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    private val eventListener =
        EventListener { value: QuerySnapshot?, error: FirebaseFirestoreException? ->
            try {
                if (error != null) {
                    return@EventListener
                }
                if (value != null) {
                    for (documentChange in value.documentChanges) {
                        if (documentChange.type == DocumentChange.Type.ADDED) {
                            val senderId =
                                documentChange.document.getString(Constants.KEY_SENDER_ID)
                            val receiverId =
                                documentChange.document.getString(Constants.KEY_RECEIVER_ID)
                            val chatMessage = ChatMessage()
                            chatMessage.senderId = senderId
                            chatMessage.receiverId = receiverId
                            if (preferenceManager!!.getString(Constants.KEY_USER_ID) == senderId) {
                                chatMessage.conversionImage =
                                    documentChange.document.getString(Constants.KEY_RECEIVER_IMAGE)
                                chatMessage.conversionName =
                                    documentChange.document.getString(Constants.KEY_RECEIVER_NAME)
                                chatMessage.conversionId =
                                    documentChange.document.getString(Constants.KEY_RECEIVER_ID)
                            } else {
                                chatMessage.conversionImage =
                                    documentChange.document.getString(Constants.KEY_SENDER_IMAGE)
                                chatMessage.conversionName =
                                    documentChange.document.getString(Constants.KEY_SENDER_NAME)
                                chatMessage.conversionId =
                                    documentChange.document.getString(Constants.KEY_SENDER_ID)
                            }
                            chatMessage.message =
                                documentChange.document.getString(Constants.KEY_LAST_MESSAGE)
                            chatMessage.dateObject =
                                documentChange.document.getDate(Constants.KEY_TIMESTAMP)
                            conversions!!.add(chatMessage)
                        } else if (documentChange.type == DocumentChange.Type.MODIFIED) {
                            for (i in conversions!!.indices) {
                                val senderId =
                                    documentChange.document.getString(Constants.KEY_SENDER_ID)
                                val receiverId =
                                    documentChange.document.getString(Constants.KEY_RECEIVER_ID)
                                if (conversions!![i].senderId == senderId && conversions!![i].receiverId == receiverId) {
                                    conversions!![i].message =
                                        documentChange.document.getString(Constants.KEY_LAST_MESSAGE)
                                    conversions!![i].dateObject =
                                        documentChange.document.getDate(Constants.KEY_TIMESTAMP)
                                    break
                                }
                            }
                        }
                    }
                    Collections.sort(conversions) { obj1: ChatMessage, obj2: ChatMessage ->
                        obj2.dateObject!!.compareTo(
                            obj1.dateObject
                        )
                    }
                    conversionAdapter!!.notifyDataSetChanged()
                    binding!!.conversionRecyclerView.smoothScrollToPosition(0)
                    binding!!.conversionRecyclerView.visibility = View.VISIBLE
                    binding!!.progressBar.visibility = View.GONE
                }
            } catch (e: Exception) {
            }
        }
    private val token: Unit
        private get() {
            try {
                FirebaseMessaging.getInstance().token.addOnSuccessListener { token: String ->
                    updateToken(
                        token
                    )
                }
            } catch (e: Exception) {
            }
        }

    private fun updateToken(token: String) {
        try {
            preferenceManager!!.putString(Constants.KEY_FCM_TOKEN, token)
            val database = FirebaseFirestore.getInstance()
            val documentReference = database.collection(Constants.KEY_COLLECTION_USERS).document(
                preferenceManager!!.getString(Constants.KEY_USER_ID)!!
            )
            documentReference.update(Constants.KEY_FCM_TOKEN, token)
                .addOnFailureListener { e: Exception? -> showToast("Unable To Update Token") }
        } catch (e: Exception) {
        }
    }

    private fun logOut() {
        try {
            showToast("LogOut...") // Assuming this is a method to show a toast message
            val database = FirebaseFirestore.getInstance()
            val documentReference = database.collection(Constants.KEY_COLLECTION_USERS).document(
                preferenceManager!!.getString(Constants.KEY_USER_ID)!!
            )
            val update = HashMap<String, Any>()
            update[Constants.KEY_FCM_TOKEN] = FieldValue.delete()
            documentReference.update(update).addOnSuccessListener { unused: Void? ->
                // Clear preferences
                preferenceManager!!.clear()
                startActivity(Intent(applicationContext, LogInActivity::class.java))
                finish()
            }.addOnFailureListener { showToast("Faiuld LogOut") }
        } catch (e: Exception) {
            // Handle exceptions
        }
    }

    override fun OnConversionClicked(user: User?) {
        try {
            val intent = Intent(applicationContext, ChatActivity::class.java)
            intent.putExtra(Constants.KEY_USER, user)
            startActivity(intent)
        } catch (e: Exception) {
        }
    }

    private fun showNotification() {
        try {
            val customSoundUri =
                Uri.parse("android.resource://" + packageName + "/" + R.raw.notification)
            val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager

            // Since android Oreo notification channel is needed.
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {

                // Set the custom sound for the notification.
                val audioAttributes = AudioAttributes.Builder()
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .setUsage(AudioAttributes.USAGE_NOTIFICATION).build()
                val channel =
                    NotificationChannel(channelId, channelname, NotificationManager.IMPORTANCE_HIGH)
                channel.setSound(customSoundUri, audioAttributes)
                channel.setShowBadge(true)
                val vibrationPattern = longArrayOf(100, 50, 100, 50, 100, 50, 100, 50, 100)
                channel.vibrationPattern = vibrationPattern
                notificationManager.createNotificationChannel(channel)
            }
        } catch (e: Exception) {
        }
    }

    private fun permissionNotification() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            val permissionNotif = ActivityCompat.checkSelfPermission(
                this@MainActivity,
                Manifest.permission.POST_NOTIFICATIONS
            )
            if (permissionNotif != PackageManager.PERMISSION_GRANTED) {
                val NOTIF_PERM = arrayOf(Manifest.permission.POST_NOTIFICATIONS)
                ActivityCompat.requestPermissions(this, NOTIF_PERM, 10)
            }
        }
    }

    override fun onStart() {
        super.onStart()
        try {
            loadUserDetails()
            token
            permissionNotification()
            showNotification()
        } catch (e: Exception) {
        }
    }
}